homeassistant/components/rainbird/__init__.py

```diff
@@ -20,24 +25,16 @@
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up the config entry for Rain Bird."""
 
-    hass.data.setdefault(DOMAIN, {})
-
     controller = AsyncRainbirdController(
         AsyncRainbirdClient(
             async_get_clientsession(hass),
             entry.data[CONF_HOST],
             entry.data[CONF_PASSWORD],
         )
     )
-    coordinator = RainbirdUpdateCoordinator(
-        hass,
-        name=entry.title,
-        controller=controller,
-        serial_number=entry.data[CONF_SERIAL_NUMBER],
-    )
-    await coordinator.async_config_entry_first_refresh()
-
-    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = coordinator
+    data = RainbirdData(hass, entry, controller)
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = data
+    await data.coordinator.async_config_entry_first_refresh()
```

Comment:
Usually we don't store the data in `hass.data` until after the first successful refresh